sql-plugin/src/main/scala/com/nvidia/spark/rapids/lore/GpuLore.scala

```diff
@@ -0,0 +1,303 @@
+                    tagForSubqueryPlan(1, f.condition, loreOutputInfo)
+                  case p: GpuProjectExec =>
+                    p.projectList.foldLeft(1)((nextPlanId, expr) =>
+                      tagForSubqueryPlan(nextPlanId, expr, loreOutputInfo))
+                  case agg: GpuHashAggregateExec =>
+                    agg.aggregateExpressions.flatMap(_.aggregateFunction.children).collect {
+                      // The reason we can't support dumping subquery expression in aggregate
+                      // function is that typically aggregation function will be split into
+                      // partial aggregation and final aggregation, and the
+                      // [ReuseExchangeAndSubquery] rule will replace final aggregation's subquery
+                      // with reused subquery expression. The problem is this rule happens in
+                      // last step, even after columnar rule, so the tag will no longer work. We
+                      // may add some physical rules to handle this in future, but given that
+                      // this is a corner case, we don't support it for now.
+                        case _: ExecSubqueryExpression => throw new IllegalArgumentException(
+                          "Unable to support dumping subquery expression in aggregate function")
+                      }
+                  case _ =>
+                }
+              }
+            }
+          case _ =>
+        }
+
+        sparkPlan
+      case None =>
+        // We don't need to dump the output of the nodes, just tag the lore id
+        sparkPlan.foreachUp {
+          case g: GpuExec =>
+            nextLoreIdOf(g).foreach { loreId =>
+              g.setTagValue(LORE_ID_TAG, loreId.toString)
+            }
+          case _ =>
+        }
+
+        sparkPlan
+    }
+
+    newPlan
+  }
+
+  def loreIdOf(node: SparkPlan): Option[String] = {
+    node.getTagValue(LORE_ID_TAG)
+  }
+
+  private def tagForSubqueryPlan(startId: Int, expression: Expression,
+      loreOutputInfo: LoreOutputInfo): Int = {
+    var nextPlanId = startId
+    expression.foreachUp {
+      case sub: ExecSubqueryExpression =>
+        val innerPlan = sub.plan.child
+        if (innerPlan.isInstanceOf[GpuExec]) {
+          val dumpRDDInfo = LoreDumpRDDInfo(nextPlanId, loreOutputInfo, innerPlan.output)
+          innerPlan match {
+            case p: GpuColumnarToRowExec => p.child.setTagValue(LORE_DUMP_RDD_TAG, dumpRDDInfo)
+            case c => c.setTagValue(LORE_DUMP_RDD_TAG, dumpRDDInfo)
+          }
+
+          nextPlanId += 1
+        } else {
+          throw new IllegalArgumentException(s"Subquery plan ${innerPlan.getClass.getSimpleName} " +
+            s"is not a GpuExec")
+        }
+      case _ =>
+    }
+    nextPlanId
+  }
+
+  private def checkUnsupportedOperator(plan: SparkPlan): Unit = {
+    if (plan.isInstanceOf[ShimLeafExecNode] ||
+      plan.isInstanceOf[GpuCustomShuffleReaderExec]
```

Comment:
I don't know of any plans offhand without children that aren't also LeafExec, but it's not a guarantee there couldn't (accidentally?) be one in the future.  This is a defensive coding change.  We're checking for what we really need, rather than a proxy of it that could be wrong in this context.